src/bitmap/backend/atomic_bitmap_mmap.rs

```diff
@@ -0,0 +1,608 @@
+
+        let first_page = page_number(offset);
+        let last_page = page_number(offset.saturating_add(len - 1));
+        for page in first_page..=last_page {
+            if page >= self.number_of_pages {
+                break; // ignore out of bound access
+            }
+
+            // get the absolute page number
+            let page = self.pages_before_region + page;
+            self.logmem[page_block(page)].fetch_or(1 << page_bit(page), Ordering::SeqCst);
+        }
+    }
+
+    // Check whether the specified offset is marked as dirty. The `offset` is relative
+    // to the mery region so and offset of `0` references the start of the memory region.
+    // Any attempt to access beyond the end of the bitmap are simply ignored.
+    fn dirty_at(&self, offset: usize) -> bool {
+        let page = page_number(offset);
+        if page >= self.number_of_pages {
+            return false; // ignore out of bound access
+        }
+
+        // get the absolute page number
+        let page = self.pages_before_region + page;
+        let page_bit = self.logmem[page_block(page)].load(Ordering::SeqCst) & (1 << page_bit(page));
+        page_bit != 0
+    }
+}
+
+impl BitmapMmap for AtomicBitmapMmap {
+    /// Creates a new memory-mapped bitmap for the region of size `region_len` starting at
+    /// the `region_start_addr` memory address. This bitmap must fit within the mapped
+    /// memory of size `size` bytes starting at offset `offset` in the file referred to by
+    /// the file descriptor `fd`.
+    fn from_file(
+        region_start_addr: GuestAddress,
+        region_len: GuestUsize,
+        fd: BorrowedFd,
+        offset: u64,
+        len: u64,
+    ) -> io::Result<Self> {
+        let region_start_addr: usize = try_into(region_start_addr.raw_value())?;
+        let region_len: usize = try_into(region_len)?;
+        let offset: isize = try_into(offset)?;
+        let len: usize = try_into(len)?;
+
+        // The size of the log should be large enough to cover all known guest addresses.
+        let region_end_addr = region_start_addr
+            .checked_add(region_len - 1)
+            .ok_or(io::Error::from(io::ErrorKind::InvalidData))?;
+        let region_end_log_block = addr_block(region_end_addr);
+        if region_end_log_block >= len {
+            return Err(io::Error::from(io::ErrorKind::InvalidData));
+        }
+
+        // Note: We could try to adjust the mapping area to only cover the memory region, but
+        // the region's starting address is not guarantee to be LOG_BLOCK_SIZE-page aligned
+        // which makes the implementation needlessly cumbersome.
+        // Note: The specification does not define whether the offset must be page-aligned or not.
+        // But, since we are receiving the offset from the frontend to be used to call mmap,
+        // we assume it is properly aligned (currently, qemu always send a 0 offset).
+        let logmem = MmapLogReg::from_file(fd, offset, len)?;
```

Comment:
If qemu sends unaligned offsets mmap will fail. I can round the offset/len to page (I did it in one of my draft), but since I need to account that the region is not "word"-aligned I don't think it's worth it